jobs/async-upload/job/download.py

```diff
@@ -74,6 +95,79 @@ def download_from_hf(uri: str, dest_dir: str) -> str:
     )
 
 
+def download_from_http(uri: str, dest_dir: str) -> str:
+    """
+    adapted from kserve:
+    https://github.com/kserve/kserve/blob/4edbb36c520c2e880842229bfc56b7f11d766822/python/storage/kserve_storage/kserve_storage.py#L698-L771
+    """
+    url = urlparse(uri)
+    filename = os.path.basename(url.path)
+    # Determine if the symbol '?' exists in the path
+    if mimetypes.guess_type(url.path)[0] is None and url.query != "":
+        mimetype, encoding = mimetypes.guess_type(url.query)
+    else:
+        mimetype, encoding = mimetypes.guess_type(url.path)
+
+    if filename == "":
+        raise ValueError(f"No filename contained in URI: {uri}")
+
+    # Get header information from host url
+    host_uri = url.hostname
+    headers_env_name = host_uri + HEADERS_SUFFIX
+    if headers_env_name in os.environ:
+        headers = json.loads(os.environ[headers_env_name])
+    else:
+        headers = {}
+
+    # Use body content workflow to defer body download until response.raw is called
+    # https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow
+    with requests.get(uri, stream=True, headers=headers) as response:
+        response.raise_for_status()
+
+        is_archive = True
+        match mimetype:
+            case "application/zip":
+                valid_content_types = ZIP_CONTENT_TYPES
+            case "application/x-tar":
+                valid_content_types = TAR_CONTENT_TYPES
+            case _:
+                valid_content_types = REGULAR_FILE_CONTENT_TYPES
+                is_archive = False
+        if not response.headers.get("Content-Type", "").startswith(valid_content_types):
+            content_types_str = ", ".join(valid_content_types)
+            raise RuntimeError(
+                f"URI {uri} appears to have MIME type {mimetype} but did not respond with any of following for 'Content-Type': {content_types_str}"
+            )
+
+        dest_path = os.path.join(dest_dir, filename)
+        os.makedirs(dest_dir, exist_ok=True)
+        with open(dest_path, "wb") as dest_file:
+            shutil.copyfileobj(response.raw, dest_file)
+
+    if is_archive:
+        dest_dir = _unpack_archive_file(dest_path, mimetype, dest_dir)
```

Comment:
Would we ever want the option to tell this function to simply download archives without extracting them?